mysql-test/main/distinct.test

```diff
@@ -928,4 +928,47 @@ drop table t1;
 
 --echo #
 --echo # end of 10.5 tests
+
+--echo #
+--echo # MDEV-30660 COUNT DISTINCT seems unnecessarily slow when run on a PK
+--echo #
+
+CREATE TABLE t1 (a INT NOT NULL PRIMARY KEY, b INT NOT NULL);
+INSERT INTO t1 VALUES (1,1), (2,1), (3,1);
+
+SELECT COUNT(DISTINCT a) FROM t1;
+SELECT COUNT(DISTINCT (SELECT a)) FROM t1;
+SELECT AVG(DISTINCT a) FROM t1;
+SELECT SUM(DISTINCT a) FROM t1;
+SELECT SUM(DISTINCT a) FROM t1 WHERE a > 1;
+SELECT SUM(DISTINCT a + b) FROM t1;
+SELECT COUNT(DISTINCT a, b) FROM t1;
+SELECT COUNT(DISTINCT b, a) FROM t1;
+SELECT COUNT(DISTINCT a, a + b) FROM t1;
+SELECT AVG(DISTINCT b + a) FROM t1;
+
+CREATE TABLE t2 (a INT);
+INSERT INTO t2 VALUES (1), (2);
+
+--echo # More than one table
+SELECT COUNT(DISTINCT t1.a) FROM t1, t2;
+SELECT AVG(DISTINCT t1.a) FROM t1, t2;
+
+--echo # Const tables
+SELECT COUNT(DISTINCT a) FROM t1, (SELECT 1) AS t2;
+SELECT AVG(DISTINCT t1.a) FROM (SELECT 1 AS a) AS t2, t1, (SELECT 2 AS a) AS t3;
+SELECT SUM(DISTINCT b) FROM t1, (SELECT 1) AS t2;
+SELECT COUNT(DISTINCT a) FROM t1, (SELECT 1 UNION SELECT 2) AS t2;
+
+--echo # Unique index on two columns
+CREATE TABLE t3 (a INT NOT NULL, b INT NOT NULL);
+INSERT INTO t3 VALUES (1,1), (2,2), (3,3);
+CREATE UNIQUE INDEX t3_a_b ON t3 (a, b);
+SELECT COUNT(DISTINCT a, b) FROM t3;
+SELECT COUNT(DISTINCT b, a) FROM t3;
+SELECT COUNT(DISTINCT b, a) FROM t3 WHERE a < 3;
```

Comment:
Would `SELECT COUNT(DISTINCT b + a) FROM t3;`  be valid for deterministic functions like `+` where arguments are FIELD_ITEM, in the index, or CONST?

Reply:
I don't think so, for example we have `UNIQUE KEY(a, b)` and values (1,2) and (2,1). `SELECT COUNT(DISTINCT b + a)` must return 1 but will return 2 if we apply SIMPLE_AGGREGATOR. This could work for something simpler using only one field: `SELECT COUNT(DISTINCT a + <const>)`. At the same time, it will not work for `SELECT COUNT(DISTINCT a + b + <const>)`. I also don't know what functions are safe to use, even the arithmetic division is questionable... 
I'm not sure we should dive deep in such particular cases unless we can generalize them somehow.

Reply:
ah right. Thanks